org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/text/correction/SourceAssistProcessor.java

```diff
@@ -138,8 +140,33 @@ public SourceAssistProcessor(PreferenceManager preferenceManager) {
 
 		// Generate hashCode() and equals()
 		if (supportsHashCodeEquals(context, type, monitor)) {
-			Optional<Either<Command, CodeAction>> hashCodeEquals = getHashCodeEqualsAction(params);
-			addSourceActionCommand($, params.getContext(), hashCodeEquals);
+			// Generate QuickAssist
+			Optional<Either<Command, CodeAction>> quickAssistHashCodeEquals = Optional.empty();
+			ASTNode node = context.getCoveredNode();
+			if (node == null) {
+				node = context.getCoveringNode();
+			}
+			if (node instanceof SimpleName) {
+				ASTNode parent = node.getParent();
+				if (parent instanceof TypeDeclaration) {

```

Comment:
`AbstractTypeDeclaration` can support enum, record, annotation type as well, why not to use it here?

Reply:
I'm not sure whether we should have this source action for `enum`, `record` and `annotation type`. 

For `enum`, `java.lang.Enum` declares `hashCode()` and `equals()` as the final method. I just tried in IDEA, for `enum` and `annotation type` it doesn't offer such code action.